[https://github.com/eclipse/xtext-eclipse/issues/1221] Serializer.

miklossy opened the pull request:
- Improve the serializeReplacement implementation by modifying the
ReplaceRegion length calculation so that it takes not only the
ICompositeNode.getTotalLength() into account, but considers if the new
text contains additional whitespaces and the old text is also followed
by white spaces, than the whitespaces contained by the original document
is also consumed by the quickfix.

Signed-off-by: Tamas Miklossy <miklossy@itemis.de>

szarnekow commented on org.eclipse.xtext/src/org/eclipse/xtext/serializer/impl/Serializer.java:
```diff
@@ -210,7 +211,28 @@ public ReplaceRegion serializeReplacement(EObject obj, SaveOptions options) {
+		return new ReplaceRegion(node.getTotalOffset(), replaceRegionLength, text);
 	}
 
+	private int calculateReplaceRegionLength(ICompositeNode node, String text) {
```
Unfortunately this is not generally applicable, e.g. languages with synthetic tokens or semantic whitespace will break. This method must at least be protected

miklossy replied:
Fixed.

miklossy replied:
> Unfortunately this is not generally applicable, e.g. languages with synthetic tokens or semantic whitespace will break.

Actually I think that it is really too dangerous to modify the code in the `xtext-core` code base. Hence it is quite difficult to provide a general solution there, I would prefer to fix the problem only on the Domainmodel example side. We can still consider a general solution later on, if lot of "client projects" claims about the problem, but for the time being a fix only for the Domainmodel example would satisfy my needs.

@cdietrich @szarnekow : If you agree, I will close this PR and provide another one in `xtext-eclipse`.

szarnekow replied:
@miklossy I didn't want to scare you and I'm sure that you'll get this straight. If you don't mind, I'll be happy to give advice and provide feedback.

szarnekow commented on org.eclipse.xtext/src/org/eclipse/xtext/serializer/impl/Serializer.java:
```diff
@@ -210,7 +211,28 @@ public ReplaceRegion serializeReplacement(EObject obj, SaveOptions options) {
+	}
+
+	private boolean whiteSpaceFollows(ICompositeNode node) {
+		return node.hasNextSibling() && node.getNextSibling() instanceof HiddenLeafNode;
```
Checks against implementation classes of the node model must be avoided. Please use `instanceof ILeafNode && node.isHidden()` instead.

szarnekow replied:
In any case, the logic here is unfortunately flawed. If I understand the intention behind this code correctly, it must be something like `node.getLeadNodes().head().isHidden()` - though the name `whiteSpaceFollows` is misleading given that comment nodes are also hidden.

miklossy replied:
Fixed.

szarnekow commented on org.eclipse.xtext/src/org/eclipse/xtext/serializer/impl/Serializer.java:
```diff
@@ -210,7 +211,28 @@ public ReplaceRegion serializeReplacement(EObject obj, SaveOptions options) {
+		return node.hasNextSibling() && node.getNextSibling() instanceof HiddenLeafNode;
+	}
+
+	private boolean containsOnlyWhiteSpaces(String text) {
```
Could be `CharMatcher.whitespac().matchesAllOf(text)`

miklossy replied:
Fixed.

szarnekow commented on org.eclipse.xtext/src/org/eclipse/xtext/serializer/impl/Serializer.java:
```diff
@@ -210,7 +211,28 @@ public ReplaceRegion serializeReplacement(EObject obj, SaveOptions options) {
+		return oldTextLength;
+	}
+
+	private boolean whiteSpaceFollows(ICompositeNode node) {
```
whiteSpace -> whitespace

miklossy replied:
Fixed.

szarnekow commented on org.eclipse.xtext/src/org/eclipse/xtext/serializer/impl/Serializer.java:
```diff
@@ -210,7 +211,28 @@ public ReplaceRegion serializeReplacement(EObject obj, SaveOptions options) {
+		return node.hasNextSibling() && node.getNextSibling() instanceof HiddenLeafNode;
+	}
+
+	private boolean containsOnlyWhiteSpaces(String text) {
```
containsOnlyWhiteSpaces -> isWhitespace

miklossy replied:
Fixed.

miklossy commented:
@szarnekow I updated the PR as requested: I added a SerializerReplacementCalculationTest test case to the xtext-core repository based on the  [NoJdtTestLanguage](https://github.com/eclipse/xtext-core/blob/master/org.eclipse.xtext.testlanguages/src/org/eclipse/xtext/testlanguages/noJdt/NoJdtTestLanguage.xtext).

szarnekow commented on org.eclipse.xtext/src/org/eclipse/xtext/serializer/impl/Serializer.java:
```diff
@@ -210,7 +213,35 @@ public ReplaceRegion serializeReplacement(EObject obj, SaveOptions options) {
+	protected boolean hiddenNodeFollows(ICompositeNode node) {
+		if (node.hasNextSibling()) {
+			INode nextSibling = node.getNextSibling();
+			if (nextSibling instanceof ILeafNode) {
```
If the next sibling is a composite node, it's first child may still be a hidden node.
If you want to catch those, too, I recommend to use something like nextSibling.getLeafNodes() and check the first node in that list.

miklossy commented:
Thanks for the feedback @szarnekow . I updated the PR as requested.

> There may be corner cases where the next sibling does not exist but the next node is still a hidden node (if current is the last child in its parent).
The current fix is actually strong enough to catch two different use cases (`Domainmodel example` and `NoJdtTestLanguage`), so I considered this fix already as an improvement, even if it does not cover all use cases. I would consider the corner cases if we can find a use case that is not covered by the current fix, so we can translate that use case into a test case and improve the implementation based on that. But that I would consider as future work in a separate issue.

szarnekow commented:
I assume that you are fully aware of the fact that Xtext is neither a framework that helps to run the domainmodel example nor a framework to build a painfully trivial test language.
If every bugfix was done under the proposition of ignoring corner cases, the framework would not be where it was now.

miklossy commented:
> There may be corner cases where the next sibling does not exist but the next node is still a hidden node (if current is the last child in its parent).

Is there already any `getNextNode()` method that I can use for this purpose?

szarnekow commented on org.eclipse.xtext/src/org/eclipse/xtext/serializer/impl/Serializer.java:
```diff
@@ -210,7 +213,50 @@ public ReplaceRegion serializeReplacement(EObject obj, SaveOptions options) {
+		return new ReplaceRegion(node.getTotalOffset(), replaceRegionLength, text);
 	}
 
+	protected int calculateReplaceRegionLength(ICompositeNode node, String text) {
```
Please also add `@Since` tags to the new methods.

miklossy commented:
`@since 2.22` tags have been added.

szarnekow commented on org.eclipse.xtext/src/org/eclipse/xtext/serializer/impl/Serializer.java:
```diff
@@ -210,7 +213,60 @@ public ReplaceRegion serializeReplacement(EObject obj, SaveOptions options) {
+	/**
+	 * @since 2.22
+	 */
+	private boolean isWhitespace(String text) {
```
Last thing that slipped through: This method should either be protected or the since tag is unnecessary.

miklossy replied:
I changed the method visibility to `protected`.
